src/main/java/org/folio/circulation/resources/CirculationRulesResource.java

```diff
@@ -89,19 +107,32 @@ private void get(RoutingContext routingContext) {
   @SuppressWarnings("squid:S2147")
   private void put(RoutingContext routingContext) {
     final Clients clients = Clients.create(new WebContext(routingContext), client);
-    CollectionResourceClient loansRulesClient = clients.circulationRulesStorage();
 
-    if (loansRulesClient == null) {
+    if (clients.circulationRulesStorage() == null) {
       internalError(routingContext.response(),
         "Cannot initialise client to storage interface");
       return;
     }
 
+    getExistingPolicyIds(clients)
+      .thenAccept(result -> proceedWithUpdate(result.value(), routingContext, clients));
+  }
+
+  private void proceedWithUpdate(
+    Map<String, Set<String>> existingPoliciesIds, RoutingContext routingContext, Clients clients) {
+
     JsonObject rulesInput;
     try {
       // try to convert, do not safe if conversion fails
       rulesInput = routingContext.getBodyAsJson();
-      Text2Drools.convert(rulesInput.getString("rulesAsText"));
+      Text2Drools.convert(rulesInput.getString("rulesAsText"),
+        (policyType, policies, token) -> {
+          if (!isPolicyExisting(existingPoliciesIds, policyType, policies)) {
```

Comment:
I think `policyExists` could be a clearer name than `isPolicyExisting`

```suggestion
          if (!policyExists(existingPoliciesIds, policyType, policies)) {
```

Reply:
renamed